xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h

```diff
@@ -294,6 +294,94 @@ class CSettingValue
 namespace kodi {
 namespace addon {
 
+/*
+ * Internally used helper class to manage processing of a "C" structure in "CPP"
+ * class.
+ *
+ * At constant, the "C" structure is copied, otherwise the given pointer is
+ * superseded and is changeable.
+ *
+ * -----------------------------------------------------------------------------
+ *
+ * Example:
+ *
+ * ~~~~~~~~~~~~~{.cpp}
+ * extern "C" typedef struct C_SAMPLE_DATA
+ * {
+ *   unsigned int iUniqueId;
+ * } C_SAMPLE_DATA;
+ *
+ * class CPPSampleData : public CStructHdl<CPPSampleData, C_SAMPLE_DATA>
+ * {
+ * public:
+ *   CPPSampleData() = default;
+ *   CPPSampleData(const CPPSampleData& sample) : CStructHdl(sample) { }
+ *   CPPSampleData(const C_SAMPLE_DATA* sample) : CStructHdl(sample) { }
+ *   CPPSampleData(C_SAMPLE_DATA* sample) : CStructHdl(sample) { }
+ *
+ *   void SetUniqueId(unsigned int uniqueId) { m_cStructure->iUniqueId = uniqueId; }
+ *   unsigned int GetUniqueId() const { return m_cStructure->iUniqueId; }
+ * };
+ *
+ * ~~~~~~~~~~~~~
+ *
+ * The class automatically works with a "C" = "CPP" or "CPP" = "C".
+ */
+template <class CPP_CLASS, typename C_STRUCT>
+class CStructHdl
+{
+public:
+  CStructHdl()
+    : m_owner(true),
+      m_cStructure(new C_STRUCT)
+  {
+  }
+
+  CStructHdl(const CPP_CLASS& cppClass)
+    : m_owner(true),
+      m_cStructure(new C_STRUCT(*cppClass.m_cStructure))
+  {
+  }
+
+  CStructHdl(const C_STRUCT* cStructure)
+    : m_owner(true),
+      m_cStructure(new C_STRUCT({*cStructure}))
+  {
+  }
+
+  CStructHdl(C_STRUCT* cStructure)
+    : m_owner(false),
+      m_cStructure(cStructure)
+  {
+  }
+
+  const CPP_CLASS& operator =(const CPP_CLASS& right)
+  {
+    m_owner = true;
+    m_cStructure = new C_STRUCT(*right.m_cStructure);
+    return *this;
+  }
+
+  const CPP_CLASS& operator =(const C_STRUCT& right)
+  {
+    m_owner = true;
+    m_cStructure = new C_STRUCT(*right);
+    return *this;
+  }
+
+  virtual ~CStructHdl()
+  {
+    if (m_owner)
+      delete m_cStructure;
+  }
+
+  operator C_STRUCT*() { return m_cStructure; }
+
+protected:
```

Comment:
Why protected, not private? Do you expect derived classes?

Reply:
With private, can child not access the structure, see: https://github.com/xbmc/xbmc/blob/caf5e7e3c39ee8337f3cec182b2748031e72d134/xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/PVRClient.h#L588

Reply:
child means derived class?

Reply:
Exact, as on the example link.

Reply:
Okay, then.